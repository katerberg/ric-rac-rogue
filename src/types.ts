import {Board} from './classes/Board';

export type Coordinate = `${number},${number}`;
export type NumberCoordinates = {x: number; y: number};
export type Choice = 'x' | 'o';
export type SpaceStatusEffect = 'blocked' | 'forced';
export type Moves = Map<Coordinate, Choice | SpaceStatusEffect | undefined>;
export type SelectionTimes = Map<Coordinate, number>;

export enum RuleType {
  WIN_CON,
  FIRST_MOVE,
  TURN_ORDER,
}

export enum RuleWinType {
  X_IN_A_ROW,
}

export enum TurnOrderType {
  TAKE_TURNS,
  TWO_TO_ONE,
}

export type Rule = {
  type: RuleType;
  turnOrderType?: TurnOrderType;
  winType?: RuleWinType;
  firstPlayer?: Choice;
  xInARow?: number;
};
export type State = {
  board: Board;
  requiredWin: number;
  currentPlayer: Choice;
  maxDepth: number;
};

export type Stats = {
  totalMoves: number;
  totalTies: number;
  totalWins: number;
  totalLosses: number;
  powerUpsActivated: number;
};

export type TerminalStatus = {
  isTerminal: boolean;
  isCat: boolean;
  isWinner: boolean;
  winner: Choice | null;
};
export enum MenuOption {
  CANCEL,
  END_RUN,
}

export enum PowerUpType {
  EXTRA_TURN,
  FLIP_TILE,
  RESET_COOLDOWN,
  COPY_COLUMN,
  COPY_ROW,
  REMOVE_COLUMN,
  REMOVE_ROW,
  INCREASE_ENERGY,
  INCREASE_MAX_ENERGY,
  INCREASE_REQUIRED_WIN,
  DECREASE_REQUIRED_WIN,
  TELEPORT_RANDOM,
  FORCE_RANDOM,
  FORCE_SPACE,
  BLOCKED_SPACE,
}

export enum StatusEffectType {
  EXTRA_TURN,
  FORCE_RANDOM,
  FORCE_SPACE,
  BLOCKED_SPACE,
}

/* eslint-disable-next-line @typescript-eslint/ban-types */
export function randomEnum<T extends Object>(anEnum: T): T[keyof T] {
  const enumValues = Object.keys(anEnum) as Array<keyof T>;
  const randomIndex = Math.floor(Math.floor(enumValues.length / 2) + Math.random() * Math.floor(enumValues.length / 2));
  const randomEnumKey = enumValues[randomIndex];
  return anEnum[randomEnumKey];
}
