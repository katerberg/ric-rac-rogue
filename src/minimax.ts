import {Board} from './classes/Board';
import {numberCoordsToCoords} from './coordinatesHelper';
import {Moves, NumberCoordinates, State, TurnOrderType} from './types';
import {checkTerminal, getTotalScore} from './winCalculation';

const transpositionTable: Record<string, {bestScore: number; bestMove: NumberCoordinates}> = {};

export function boardToTranspositionTableKey(selections: Moves, columns: number, rows: number): string {
  let key = '';
  for (let y = 0; y < columns; y++) {
    for (let x = 0; x < rows; x++) {
      key += selections.get(`${x},${y}`) ?? '_';
    }
    key += ',';
  }
  key = key.slice(0, -1);
  return key;
}

export function boardToTranspositionTableKeys(selections: Moves, columns: number, rows: number): string[] {
  let key = '';
  let keyInverted = '';
  for (let y = 0; y < columns; y++) {
    for (let x = 0; x < rows; x++) {
      key += selections.get(`${x},${y}`) ?? '_';
      keyInverted += selections.get(`${y},${x}`) ?? '_';
    }
    key += ',';
    keyInverted += ',';
  }
  key = key.slice(0, -1);
  keyInverted = keyInverted.slice(0, -1);
  return [key, keyInverted];
}

export function getBestMove(
  state: State,
  pruning: boolean,
  turnOrderType: TurnOrderType,
  previousMaximizing?: boolean,
  isMaximizing?: boolean,
  depth = 0,
  alpha = -1_000_000,
  beta = 1_000_000,
): {bestScore: number; bestMove: NumberCoordinates} {
  if (depth === 0) {
    const props = Object.getOwnPropertyNames(transpositionTable);
    for (let i = 0; i < props.length; i++) {
      delete transpositionTable[props[i]];
    }
  }
  const maximizing = isMaximizing !== undefined ? isMaximizing : state.currentPlayer === 'x';
  const availableMoves = state.board.getAvailableMoves();
  let [bestMove] = availableMoves;
  if (depth > state.maxDepth) {
    return {
      bestScore:
        getTotalScore(state.board.selections, state.board.columns, state.board.rows, state.requiredWin) +
        1000 * (maximizing ? -1 : 1),
      bestMove,
    };
  }
  //Initialize best to the worst possible value and have a default move
  const WORST_POSSIBLE_SCORE = maximizing ? -1_000_000 : 1_000_000;
  // const DEPTH_MULTIPLIER = maximizing ? 0.1 : -0.1;
  const DEPTH_MULTIPLIER = 0.1;
  let bestScore = WORST_POSSIBLE_SCORE;
  let newAlpha = alpha;
  let newBeta = beta;

  for (let i = 0; i < availableMoves.length; i++) {
    const move = availableMoves[i];
    //Initialize a new board with a copy of our current state
    const child: State = {
      ...state,
      board: new Board({
        columns: state.board.columns,
        rows: state.board.rows,
        selections: new Map(state.board.selections),
      }),
      currentPlayer: maximizing ? 'o' : 'x', //irrelevant and vestigial
    };
    child.board.selections.set(numberCoordsToCoords(move), maximizing ? 'x' : 'o');
    const terminalState = checkTerminal(child.board, child.requiredWin);
    if (terminalState.isTerminal) {
      if (terminalState.isWinner) {
        // if it's terminal, get the best score minus how long it took to get there
        bestScore = WORST_POSSIBLE_SCORE * -1 - depth * DEPTH_MULTIPLIER;
        bestMove = move;
        break;
      }
      if (terminalState.isCat) {
        if ((maximizing && 0 > bestScore) || (!maximizing && 0 < bestScore)) {
          // if it's terminal, it's zero because it's better than any loss but worse than any win
          bestScore = 0;
          bestMove = move;
        }
      }
    } else {
      let nodeValue: {bestScore: number; bestMove: NumberCoordinates};
      const transpositionTableKey = boardToTranspositionTableKey(
        child.board.selections,
        child.board.columns,
        child.board.rows,
      );
      if (transpositionTable[transpositionTableKey]) {
        nodeValue = transpositionTable[transpositionTableKey];
      } else {
        let nextMaximizing = !maximizing;
        if (turnOrderType === TurnOrderType.TWO_TO_ONE && previousMaximizing !== maximizing && !maximizing) {
          nextMaximizing = maximizing;
        }
        nodeValue = getBestMove(
          child,
          pruning,
          turnOrderType,
          maximizing,
          nextMaximizing,
          depth + 1,
          newAlpha,
          newBeta,
        );
        boardToTranspositionTableKeys(child.board.selections, child.board.columns, child.board.rows).forEach((key) => {
          if (!transpositionTable[key]) {
            transpositionTable[key] = nodeValue;
          }
        });
      }
      // alpha beta pruning
      if ((maximizing && nodeValue.bestScore > bestScore) || (!maximizing && nodeValue.bestScore < bestScore)) {
        ({bestScore} = nodeValue);
        bestMove = move;
        if (pruning) {
          if (maximizing) {
            newAlpha = Math.max(bestScore, newAlpha);
          } else if (!maximizing) {
            newBeta = Math.min(bestScore, newBeta);
          }
        }
      }
    }
    if (newBeta <= newAlpha) {
      break;
    }
  }
  return {bestScore, bestMove};
}
