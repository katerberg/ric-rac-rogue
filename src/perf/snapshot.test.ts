import {getBestMove} from '../minimax';
import {State} from '../types';

function getEmptyBoardState(): State {
  return {
    columns: 3,
    rows: 3,
    requiredWin: 3,
    maxDepth: 1200,
    selections: new Map(),
    currentPlayer: 'x',
    room: {rules: []},
  };
}

describe('snapshot', () => {
  beforeEach(() => {
    jest.setTimeout(100_000);
  });
  describe('infinite depth', () => {
    it('3x3 with 3 in a row', () => {
      const input = getEmptyBoardState();

      const time = new Date().getTime();

      getBestMove(input);

      expect(new Date().getTime() - time).toMatchSnapshot();
    });

    it('4x4 with 3 in a row', () => {
      const input = {...getEmptyBoardState(), rows: 4, columns: 4};

      const time = new Date().getTime();
      getBestMove(input);

      expect(new Date().getTime() - time).toMatchSnapshot();
    });

    it('4x10 with 3 in a row', () => {
      const input = {...getEmptyBoardState(), rows: 4, columns: 10};

      const time = new Date().getTime();
      getBestMove(input);

      expect(new Date().getTime() - time).toMatchSnapshot();
    });

    it('4x20 with 3 in a row', () => {
      const input = {...getEmptyBoardState(), rows: 4, columns: 20};

      const time = new Date().getTime();
      getBestMove(input);

      expect(new Date().getTime() - time).toMatchSnapshot();
    });
  });
  describe('limited depth', () => {
    it('4x4 with 3 in a row', () => {
      const input = {...getEmptyBoardState(), rows: 4, columns: 4, maxDepth: 10};

      const time = new Date().getTime();
      getBestMove(input);

      expect(new Date().getTime() - time).toMatchSnapshot();
    });

    it('4x4 with 4 in a row', () => {
      const input = {...getEmptyBoardState(), rows: 4, columns: 4, requiredWin: 4, maxDepth: 10};

      const time = new Date().getTime();
      getBestMove(input);

      expect(new Date().getTime() - time).toMatchSnapshot();
    });
  });
});
