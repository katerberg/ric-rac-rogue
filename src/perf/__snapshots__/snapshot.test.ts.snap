// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot infinite depth 3x3 with 3 in a row 1`] = `22`;

exports[`snapshot infinite depth 4x4 with 3 in a row 1`] = `993`;

exports[`snapshot infinite depth 4x10 with 3 in a row 1`] = `7442`;

exports[`snapshot infinite depth 4x20 with 3 in a row 1`] = `24372`;

exports[`snapshot limited depth 4x4 with 3 in a row 1`] = `979`;

exports[`snapshot limited depth 4x4 with 4 in a row 1`] = `32810`;
